### PR TITLE
Setup mailer and send mail result when time out?

### PR DESCRIPTION
**Pull request by tunv**

### CODE REVIEW

File: app/views/exams/show.html.erb [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
-  <table class="table table-bordered">
-    <tr>
-      <th><%= t "question" %></th>
-      <th><%= t "your_answer" %></th>
-      <th><%= t "correct" %></th>      
-    </tr>
-    <%= render @exam.results %>
-  </table>
+  <%= form_for [@category, @exam] do |f| %>
+  <%= render "shared/error_messages", object: f.object %>
```
Comment (phonghenry14) [inline app/views/exams/show.html.erb]: indent

File: app/views/result_mailer/result_exam.html.erb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<h2>This is your exam's results</h2>
```
Comment (phonghenry14) [inline app/views/result_mailer/result_exam.html.erb]: i18n

File: app/views/result_mailer/result_exam.html.erb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+<h3>Mark: <%= @exam.mark %></h3>
```
Comment (phonghenry14) [inline app/views/result_mailer/result_exam.html.erb]: i18n

File: app/views/result_mailer/result_exam.html.erb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+<table border="1">
+  <tr>
+    <th>Question</th>
```
Comment (phonghenry14) [inline app/views/result_mailer/result_exam.html.erb]: i18n 3 dòng

File: app/mailers/result_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class ResultMailer < ApplicationMailer
+  def result_exam exam
+    @exam = exam
+    mail(to: exam.user.email, subject: "Result Email")
```
Comment (yuyuvn) [inline app/mailers/result_mailer.rb]: remove ()

File: app/views/result_mailer/result_exam.text.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+This is your exam's results
```
Comment (phonghenry14) [inline app/views/result_mailer/result_exam.text.erb]: i18n

File: config/environments/development.rb [hunk 7]
```diff
@@ -38,4 +38,16 @@
...
   # Raises error for missing translations
   # config.action_view.raise_on_missing_translations = true
+
+  config.action_mailer.default_url_options = {host: "localhost:3000"}
+  config.action_mailer.delivery_method = :smtp
+# SMTP settings for gmail
+  config.action_mailer.smtp_settings = {
+   :address              => "smtp.gmail.com",
+   :port                 => 587,
+   :user_name            => ENV['gmail_username'],
```
Comment (phonghenry14) [inline config/environments/development.rb]: " "

File: config/environments/development.rb [hunk 8]
```diff
@@ -38,4 +38,16 @@
...
+   :password             => ENV['gmail_password'],
+   :authentication       => "plain",
+  :enable_starttls_auto => true
```
Comment (phonghenry14) [inline config/environments/development.rb]: indent

File: app/controllers/exams_controller.rb [hunk 9]
```diff
@@ -25,13 +25,14 @@ def edit
...
+      - (Time.zone.now - @exam.started_at).to_i
   end
 
-   def update
+  def update
     if @exam.update_attributes update_params
       flash[:success] = t "update_success"
-      @exam.time_out? ? (redirect_to exam_path @exam) 
+      @exam.time_out? ? (ResultMailer.result_exam(@exam).deliver) && (redirect_to exam_path @exam)
         : (redirect_to edit_exam_path @exam)
```
Comment (yuyuvn) [inline app/controllers/exams_controller.rb]: don't user ? : in 2 lines
https://github.com/framgia/coding-standards/blob/master/vn/ruby/standard.md

File: app/views/result_mailer/result_exam.html.erb [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    <th>Your answer</th>
+    <th>Correct</th>
+  </tr>
+  <% @exam.results.each do |result| %>
+    <tr>
+      <td><%= result.question.content %></td>
+      <td><%= result.answer.content unless result.answer.nil? %></td>
+      <td>
+        <%= (result.answer.nil? || !result.answer.correct?) ? "False" : "True" %>
```
Comment (yuyuvn) [inline app/views/result_mailer/result_exam.html.erb]: you can use this:
result.answer.try :correct?

File: app/views/result_mailer/result_exam.text.erb [hunk 11]
```diff
@@ -0,0 +1,8 @@
...
+Mark: <%= @exam.mark %>
+<% @exam.results.each do |result| %>
+  Question: <%= result.question.content %>, 
+  Your answer: <%= result.answer.content unless result.answer.nil? %>, 
+  Correct: 
+  <%= (result.answer.nil? || !result.answer.correct?) ? "False" : "True" %>
```
Comment (yuyuvn) [inline app/views/result_mailer/result_exam.text.erb]: you can use this:
result.answer.try :correct?

File: config/environments/development.rb (hunk 8 above)
Comment (yuyuvn) [inline config/environments/development.rb]: indent

File: config/environments/development.rb [hunk 12]
```diff
@@ -38,4 +38,16 @@
 
   # Raises error for missing translations
   # config.action_view.raise_on_missing_translations = true
+
+  config.action_mailer.default_url_options = {host: "localhost:3000"}
+  config.action_mailer.delivery_method = :smtp
+# SMTP settings for gmail
```
Comment (yuyuvn) [inline config/environments/development.rb]: indent

File: config/environments/development.rb (hunk 7 above)
Comment (yuyuvn) [inline config/environments/development.rb]: convention

File: config/environments/development.rb [hunk 13]
```diff
@@ -38,4 +38,16 @@
...
+  config.action_mailer.smtp_settings = {
+   :address              => "smtp.gmail.com",
+   :port                 => 587,
+   :user_name            => ENV['gmail_username'],
+   :password             => ENV['gmail_password'],
```
Comment (yuyuvn) [inline config/environments/development.rb]: convention

File: config/environments/production.rb [hunk 14]
```diff
@@ -76,4 +76,14 @@
...
   # Do not dump schema after migrations.
   config.active_record.dump_schema_after_migration = false
+  config.action_mailer.delivery_method = :smtp
+  # SMTP settings for mailgun
+  ActionMailer::Base.smtp_settings = {
+    :port           => 587,
+    :address        => "smtp.mailgun.org",
+    :domain         => ENV['domain'],
+    :user_name      => ENV['username'],
+    :password       => ENV['password'],
```
Comment (yuyuvn) [inline config/environments/production.rb]: convention (3 dòng)

File: config/initializers/setup_mail.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
+ActionMailer::Base.delivery_method = :smtp
+ActionMailer::Base.smtp_settings = {
+:address => "smtp.gmail.com",
+:port => 587,
+:domain => "gmail.com",
+:user_name => "nguyenvantu.dhbk@gmail.com",
+:password => "nguoiyeu",
+:authentication => "plain",
+:enable_starttls_auto => true
```
Comment (yuyuvn) [inline config/initializers/setup_mail.rb]: indent

File: app/mailers/result_mailer.rb [hunk 16]
```diff
@@ -0,0 +1,6 @@
+class ResultMailer < ApplicationMailer
+  def result_exam exam
+    @exam = exam
+    mail to: exam.user.email, subject: "Result Email"
```
Comment (yuyuvn) [inline app/mailers/result_mailer.rb]: I18n

File: app/mailers/application_mailer.rb [hunk 17]
```diff
@@ -0,0 +1,3 @@
+class ApplicationMailer < ActionMailer::Base
+  default from: "from@example.com"
```
Comment (yuyuvn) [inline app/mailers/application_mailer.rb]: move to settings

File: config/environments/development.rb [hunk 18]
```diff
@@ -38,4 +38,15 @@
 
   # Raises error for missing translations
   # config.action_view.raise_on_missing_translations = true
+
+  config.action_mailer.default_url_options = {host: "localhost:3000"}
```
Comment (vuducdat) [inline config/environments/development.rb]: "localhost:3000" should input setting.

File: app/models/exam.rb [hunk 19]
```diff
@@ -22,6 +22,12 @@ def time_out?
     end
   end
 
+  def update_status
+    if time_out?
```
Comment (dieunb) [inline app/models/exam.rb]: one line of code

File: app/controllers/exams_controller.rb [hunk 20]
```diff
@@ -28,13 +28,15 @@ def edit
...
 
-   def update
-    if @exam.update_attributes update_params
-      flash[:success] = t "update_success"
-      @exam.time_out? ? (redirect_to exam_path @exam) 
-        : (redirect_to edit_exam_path @exam)
-    else
-      redirect_to :back
+  def update
+    if @exam.status == Settings.user.view
```
Comment (dieunb) [inline app/controllers/exams_controller.rb]: `Settings.user.view ==  @exam.status`

File: config/initializers/setup_mail.rb [hunk 21]
```diff
@@ -0,0 +1,10 @@
+ActionMailer::Base.delivery_method = :smtp
+ActionMailer::Base.smtp_settings = {
+  :address => "smtp.gmail.com",
+  :port => 587,
+  :domain => "gmail.com",
+  :user_name => "nguyenvantu.dhbk@gmail.com",
```
Comment (dieunb) [inline config/initializers/setup_mail.rb]: store value in environment variable

File: config/initializers/setup_mail.rb [hunk 22]
```diff
@@ -0,0 +1,10 @@
...
+  :password => "nguoiyeu",
```
Comment (dieunb) [inline config/initializers/setup_mail.rb]: store value in environment variable

File: config/initializers/setup_mail.rb [hunk 23]
```diff
@@ -0,0 +1,10 @@
+ActionMailer::Base.delivery_method = :smtp
+ActionMailer::Base.smtp_settings = {
+  :address => "smtp.gmail.com",
```
Comment (dieunb) [inline config/initializers/setup_mail.rb]: convention

**Pull request merged**